lib/logstash/codecs/cef.rb

```diff
@@ -58,28 +98,115 @@ def decode(data)
+    if severity == self.class.get_config["severity"][:default]
+      # Use deprecated setting sev
+      severity = sanitize_severity(event, @sev)
+    end
 
-    # Signature, Name, and Sev should be set in the config, with ref to fields
     # Should also probably set the fields sent
-    header = ["CEF:0", "Elasticsearch", "Logstash", "1.0", @signature, @name, @sev].join("|")
-    values = @fields.map {|name| get_value(name, data)}.join(" ")
-    # values = values.map {|k,v| "#{k}=#{v}"}.join(" ")
-    @on_event.call(header + " " + values + "\n")
+    header = ["CEF:0", vendor, product, version, signature, name, severity].join("|")
+    values = @fields.map {|fieldname| get_value(fieldname, event)}.compact.join(" ") unless fields.nil?
+
+    @on_event.call(event, header + "|" + values.to_s + "\n")
+  end
+
+  # Escape pipes and backslashes in the header. Equal signs are ok.
+  # Newlines are forbidden.
+  def sanitize_header_field(value)
+    output = ""
+
+    value = value.gsub(/\r\n/, "\n")
+
+    value.each_char{|c|
+      case c
+      when "\\", "|"
+        output += "\\" + c
+      when "\n", "\r"
+        output += " "
+      else
+        output += c
+      end
+    }
+
+    return output
+  end
+
+  # Keys must be made up of a single word, with no spaces
+  # must be alphanumeric
+  def sanitize_extension_key(value)
+    value = value.gsub(/[^a-zA-Z0-9]/, "")
+    return value
+  end
+
+  # Escape equal signs in the extensions. Canonicalize newlines.
+  # CEF spec leaves it up to us to choose \r or \n for newline.
+  # We choose \n as the default.
+  def sanitize_extension_val(value)
+    output = ""
+
+    value = value.gsub(/\r\n/, "\n")
+
+    value.each_char{|c|
+      case c
+      when "\\", "="
+        output += "\\" + c
+      when "\n", "\r"
+        output += "\\n"
+      else
+        output += c
+      end
+    }
+
+    return output
   end
 
   private
-  def get_value(name, event)
-    val = event[name]
+
+  def get_value(fieldname, event)
+    val = event[fieldname]
+
+    return nil if val.nil?
+
     case val
+    when Array
+      return sanitize_extension_key(fieldname) + "=" + sanitize_extension_val(val.to_json)
     when Hash
-      return name + "=" + val.to_json
+      return sanitize_extension_key(fieldname) + "=" + sanitize_extension_val(val.to_json)
+    when LogStash::Timestamp
+      return sanitize_extension_key(fieldname) + "=" + val.to_s
     else
-      return name + "=" + val
+      return sanitize_extension_key(fieldname) + "=" + sanitize_extension_val(val)
     end
   end
 
+  def sanitize_severity(event, severity)
+    severity = sanitize_header_field(event.sprintf(severity)).strip
+    severity = self.class.get_config["severity"][:default] unless valid_severity?(severity)
+    severity = severity.to_i.to_s
+  end
+
+  def valid_severity?(sev)
+    return (sev.to_i.to_s == sev.to_s || sev.to_f.to_s == sev.to_s && sev.to_f - sev.to_i == 0) && sev.to_i >= 0 && sev.to_i <= 10
```

Comment:
for readability, I'd suggest breaking this into multiple lines. also, can you describe a bit the conditions for severity validity?
My suggestion:

```ruby
def valid_severity?(sev)
  f = Float(sev)
  # check if it's an integer or a float with no remainder
  # and if the value is between 0 and 10 (inclusive)
  (f % 1 == 0) && f.between?(0,10)
rescue TypeError, ArgumentError
  false
end

[
  valid_severity?(nil) == false,
  valid_severity?("") == false,
  valid_severity?("foo") == false,
  valid_severity?("1.5") == false,
  valid_severity?("-1") == false,
  valid_severity?("11") == false,
  valid_severity?("0") == true,
  valid_severity?("10") == true,
  valid_severity?("1.0") == true,
  valid_severity?(1) == true,
  valid_severity?(1.0) == true
].all?

# true
```